Add content on performance benchmarks.

khushi-411 opened the pull request:
This PR adds content on performance. Follows issue: https://github.com/numpy/numpy.org/issues/370

Rendered Version: https://deploy-preview-461--numpy-preview.netlify.app/

cc: @mattip

mattip commented on content/en/config.yaml:
```diff
@@ -75,6 +75,7 @@ params:
       text: NumPy supports a wide range of hardware and computing platforms, and plays well with distributed, GPU, and sparse array libraries.
     - title: Performant
       text: The core of NumPy is well-optimized C code. Enjoy the flexibility of Python with the speed of compiled code.
+      url: /benchmark
```
Is this meant to create a link?

mattip commented:
Please reformat the lines to <80 characters

mattip commented:
How  to best link to this from the “Performant” block on the front page?

rgommers commented:
> How to best link to this from the “Performant” block on the front page?

The content is in `content/en/config.yaml`, under `keyfeatures:`. An item `url:` could be added there under `title` and `text`, and that can then be used in `layouts/partials/keyfeatures.html` (or `static/css/keyfeatures.css`, not sure without looking in more detail).

mattip commented:
Please reformat the lines in the markdown to have less than 80 characters.

khushi-411 commented:
Hi, 

I'm trying to run the following test script again and again, and I'm getting significant difference in the results. 
Script:
```
import numpy as np
import timeit

k = np.array([[0.023, 0.043, 0.024], [0.053, 0.035, 0.065]])
s = np.array([[0.045, 0.056, 0.073], [0.034, 0.075, 0.089]])

print(type(k))
print(type(k[1][2]))

def sum():
    result = k + s

def add():
    result = np.add(k, s)

def array_add():
    result = k + s

print('sum', timeit.timeit('sum()', globals = globals(), number = 100000))
print('add', timeit.timeit('add()', globals = globals(), number = 100000))
print('array_add', timeit.timeit('array_add()', globals=globals(), number = 100000))
```

The unit of the measured time is microseconds. Tabular representation: 
    |             |     sum()        |     np.add()   |     k + s        |
    |  --------  |      --------      |      --------      |     --------      |
    | 100       | 75.8810       | 72.8020       | 63.2700       |   
    | 10000   | 6438.8619   | 5902.6079   |  6048.7920  |
    | 100000 | 59495.1210 | 58891.7659 | 67185.8069 |
    | 100000 | 61379.8119 | 67006.0630 | 56408.0749 |
    | 100000 | 57788.9550 | 63652.3270 | 61401.5459 |
    | 100000 | 60107.9349 | 64737.0580 | 55648.8030 |
    | 100000 | 57868.9699 | 64138.9370 | 63911.6729 |
    | 100000 | 56889.7249 | 58800.8449 | 56531.0189 |

This problem is not only with the sumation functions, but with others too. Please have a look in the slides. The link is attached in the end.
I'm executing it on my local system using `python filename.py` .  Also, I know that these are very small changes (like in microseconds) though I am concerned about it because it is creating huge differences in our main model. Like:

* Time taken to execute the N-body problem: `78.864s` .
* When I changed `np.square(vector)` to `vector ** 2` , time taken to execute the program is `69.276s`. 
* I also changed the `np.multiply()` functions to `a * b` , then the program took: `69.276s`.

One solution I can think off is, to find variances and standard deviations of various functions on iterating over and over again. And then select the best option among them. Does that makes sense? 
Do you have any other options to so so?

Link to slides: https://hackmd.io/@qEVTAJWCQPe7LdnpHWF-CQ/r1al9Ehgt

Thanks!

cc: @mattip @rgommers

mattip commented:
Deployment is failing with
```
$ python gen_config.py && hugo
Start building sites …
Total in 114 ms
Error: Error building site: process: readAndProcessContent: \
    open /opt/build/repo/public/config.yaml: no such file 
```

mattip commented:
The results look reasonable now. Did you regenerate the figure to reflect the table?

seberg reviewed (commented):
I don't really want to dive into this, but looked at the code and found it a bit oddly dense for the NumPy version.  Is the code taken from somewhere?  I am surprised, because the nbable code seemed a bit clearer/and more comparable to me on first sight.

To be clear: I like this start and having N-Body seems good!  My code comments are mainly because it seemed the code could be better and needs to be more comparable.

But, the document around it needs much more thoughts in my opinion.  As is, I think the N-body examples probably cannot stand alone (they are just too terrible for NumPy, NumPy is not good at this, although it might be decent for large enough `N`, in which case that may work).

Even then, we need a story around it to go somewhat beyond "numpy is fast" (or "how fast is numpy").

Pythran/numba are here for the "fill the gap" story part: NumPy is fast enough for most things, and it is easy to fill the gap with C/Pythran/numba/cython where needed.  It is trivial for Julia, etc. to point out that they are faster than Python/NumPy because it plays to their strengths, lets try to play to our strengths?  But even that feels hard if we give examples 50-100 times slower than C, which is really much worse than typical).

mattip commented:
> As is, I think the N-body examples probably cannot stand alone 

This is meant to be the first of many benchmarks. We chose this because it was well constructed and seemed to be straight-forward to present. It turned out to be not so great for NumPy, although at the end of the day I think there should be a way to avoid the loops and get better results.

khushi-411 commented on benchmarks/python/optimized-numpy.py:
```diff
@@ -0,0 +1,123 @@
+        distances = np.square(vectors).sum(axis = 1)
+        coefs = distances ** 1.5
+
+        _x = np.add(np.sum(
+            np.divide(
+                np.multiply(
+                    masses[index_p0 + 1: nb_particles], -1 * vectors.T
+                    ),
+                coefs).T, axis = 0),
+            accelerations[index_p0]
+            )
```
Hi,
I agree with you that we can make the code simpler. I tried to do so, especially at this part. I observed a few things which I would like to share and get your inputs. 

I tried the following ways to calculate accelerations:

- The part I had updated in PR.
    - The time it took: `187.2880`.
- When I changed that part too:
  ```
  ind0_mul = masses[index_p0 + 1: nb_particles] * (-1 * vectors.T)
  ind0_div = ind0_mul / coefs
  ind0_div = ind0_div.T
  ind0_sum = np.sum(ind0_div, axis = 0)
  accelerations[index_p0] += accelerations[index_p0]
  
  ind1_mul = mass0 * vectors.T
  ind1_div = ind1_mul / coefs
  ind1_div = ind1_div.T
  accelerations[index_p0 + 1: nb_particles] += ind1_div
  ```
   - The time it took: `196.3231`.
- Another way I tried is the simplest to understand. But does not perform well is.
  ```
  def compute_accelerations(accelerations, masses, positions):
      nb_particles = masses.size
  
      for index_p0 in range(nb_particles - 1):
          position0 = positions[index_p0]
          mass0 = masses[index_p0]
  
          vector = np.empty(3)
  
          for index_p1 in range(index_p0 + 1, nb_particles):
              mass1 = masses[index_p1]
              position1 = positions[index_p1]
  
              for i in range(3):
                  vector[i] = position0[i] - position1[i]
              
              distance = sum(vector ** 2) * math.sqrt(sum(vector ** 2))
              coef_m1 = mass0 / distance
              coef_m2 = mass1 / distance
  
              for i in range(3):
                  accelerations[index_p0][i] += coef_m1 * -1 * vector[i]
                  accelerations[index_p1][i] += coef_m2 * vector[i]
  
      return accelerations
  ```
  - The time it took: `475.0748`.

#### Prerequisite
- These results are for `50` iterations, for datasets of input size `16` at an interval of `10` seconds, after each `0.001` step.
- **Command to run the files**:
   ```
    taskset -c 6,7,14,15 python python/optimized-numpy.py data/input16.txt 10.0
  ```
  Here, 
  - `taskset -c 6,7,14,15`: for running files in isolated cpu's.
  - `python/optimized-numpy.py`: python file, whose location is in the python folder.
  - `data/input16.txt`: denotes dataset.
  - `10.0`: total time to iterate.
- **Machine Configuration**:
  - OS: `ManjaroLinux 21.1.1`
  - Model Name: `Intel(R) Core(TM) i7-10870H CPU @ 2.20GHz`
  - RAM: `16GB`
- **Dependencies**:
  - Python: `3.9.6`
  - NumPy: `1.20.3`
  - Pythran: `0.9.12.post1`
  - Transonic: `0.4.10`
  - Numba: `0.54.0`
  - CMake: `3.21.1`
  - GCC: `11.1.0`

seberg replied:
The last version is nice, cooking it down to proper numpy seems not so bad (i.e. remove for loops, except possibly the outer one), then consider using in-place logic.

My main criticism is that it feels you do different operations compared to C++, and that changes the theoretical work done by the programs.

For example:
```
ind0_mul = masses[index_p0 + 1: nb_particles] * (-1 * vectors.T)
ind0_div = ind0_mul / coefs
```
vs.:
```
coef_m1 = mass0 / distance
``
reorders the multiplication with vectors.  The reorder itself is fine (although it would be nicer to avoid for comparison?).  But `ind0_mul` already includes the `for i in range(3):` loop (and maybe more?) so it has to do 3 times as many divisions.

I see the divisions, because they are relatively slow.  Of course this may have nothing to do with the timings you observe.  A profiler (line profiler) might help understand them easier.

Another example, is that I bet the C++ code does not do `-1 * vectors`, but rather uses `acceleration -= ...` at the end, which avoids one operation (albeit probably one that just doesn't matter).

khushi-411 replied:
Hi, 
Thanks for the brief description. As suggested earlier, I used Python operators and removed the for-loop. Also, I modified the code to maintain uniformity between them. 
Yes! Now it's more readable, and the performance of the model is also improved. 

> I see the divisions, because they are relatively slow. Of course this may have nothing to do with the timings you observe. A profiler (line profiler) might help understand them easier.

Yes, divisions are really slow. I had modified it and updated the PR. 

> Another example, is that I bet the C++ code does not do -1 * vectors, but rather uses acceleration -= ... at the end, which avoids one operation

Indeed, You won the bet!!

But  I have a different kind of opinion on this part and would like to discuss it. I agree that using `acceleration -= ...` avoids one operation that might significantly increase the performance of the model for larger datasets. At the same time,
Let us consider two bodies at a distance, `d`, under mutual gravitational force of attraction. The 1st body is moving from position `i` to position `j`. Hence, the position vector of the 1st body will be represented as `r_j - r_i`. And the 2nd body is moving from position `j` to position `i`. Its position vector is represented as `r_i - r_j`. 
Therefore I concluded that it is conceptually more strong to use `-1 * vectors` than to use `acceleration -= ...`. Though, the results are the same. Does that make sense? What do you think?

khushi-411 replied:
The following are the results of line_profiler from a few methods I tried:

- The best one (till now):
  ```
  time taken: 7.4037153270001
  Wrote profile results to optimized_numpy5.py.lprof
  Timer unit: 1e-06 s
  
  Total time: 6.60908 s
  File: python/optimized_numpy5.py
  Function: compute_accelerations at line 20
  
  Line #      Hits         Time  Per Hit   % Time  Line Contents
  ==============================================================
      20                                           @profile
      21                                           def compute_accelerations(accelerations, masses, positions):
      22     10002       7759.0      0.8      0.1      nb_particles = masses.size
      23                                           
      24    160032     103357.0      0.6      1.6      for index_p0 in range(nb_particles - 1):
      25    150030     126036.0      0.8      1.9          position0 = positions[index_p0]
      26    150030     101733.0      0.7      1.5          mass0 = masses[index_p0]
      27                                           
      28    150030     398855.0      2.7      6.0          vectors = position0 - positions[index_p0 + 1: nb_particles]
      29                                           
      30    150030    1057577.0      7.0     16.0          distances = (vectors**2).sum(axis=1)
      31    150030     468228.0      3.1      7.1          coefs = distances**1.5
      32                                           
      33    150030    3144079.0     21.0     47.6          accelerations[index_p0].__iadd__(np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T / coefs).T, axis=0))
      34    150030    1197118.0      8.0     18.1          accelerations[index_p0 + 1: nb_particles].__iadd__((mass0 * vectors.T / coefs).T)
      35                                           
      36     10002       4341.0      0.4      0.1      return accelerations
  ```

- Without using in-place logic:
  ```
  35    150030    2897902.0     19.3     46.3          accelerations[index_p0] += np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis=0)
  36    150030    1083315.0      7.2     17.3          accelerations[index_p0 + 1: nb_particles] += (mass0 * vectors.T * coefs).T
  ```

- When I changed some method to calculate distances:
  ```
  30    150030    1938345.0     12.9     27.3          distances = np.sum((vectors**2), axis=1) #.sum(axis=1)
  31    150030     601801.0      4.0      8.5          coefs = 1./distances**1.5
  ```
- Another way of the same:
  ```
  32    150030     806488.0      5.4     12.5          coefs = 1./np.sqrt(distances**3)
  33    150030    2879604.0     19.2     44.6          accelerations[index_p0].__iadd__(np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis=0))
  34    150030    1064629.0      7.1     16.5          accelerations[index_p0 + 1: nb_particles].__iadd__((mass0 * vectors.T * coefs).T)
  ```
- The most important, that I used earlier (this was really very slow, comparatively):
  ```
  31    150030     424432.0      2.8      7.1          coefs = distances**1.5
  32                                                   #coefs = 1./distances**1.5
  33                                                   #coefs = 1./np.sqrt(distances**3)
  34    150030    2849387.0     19.0     47.6          accelerations[index_p0].__iadd__(np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T / coefs).T, axis=0))
  35    150030    1070838.0      7.1     17.9          accelerations[index_p0 + 1: nb_particles].__iadd__((mass0 * vectors.T / coefs).T)
  ```

khushi-411 commented on benchmarks/python/optimized-numpy.py:
```diff
@@ -0,0 +1,123 @@
+
+        vectors = position0 - positions[index_p0 + 1: nb_particles]
+
+        distances = np.square(vectors).sum(axis = 1)
```
I'll modify it. Thanks!

khushi-411 commented on benchmarks/python/optimized-numpy.py:
```diff
@@ -0,0 +1,123 @@
+        accelerations.fill(0)
+        accelerations = compute_accelerations(accelerations, masses, positions)
+
+        velocities = 0.5 * np.multiply(time_step, np.add(accelerations, accelerations1)) + velocities
```
In this case, we are multiplying a scaler `time_step` with an array `np.add(accelerations, accelerations1)`. I created a test case to check their behavior and got the following output. The results are absurd. We are getting different results while running the same file multiple times. 

I implemented `np.multiply()` because I felt NumPy's code should consist more of NumPy functions than Python's operator. This might not be a strong reason or a valid reason. Would like to know your viewpoints?

|  | np.multiply() | scaler * array      |
| :---         |     :---:      |          :---: |
 | 100         | 77.6910         | 70.4279      |
 | 10000    | 6722.0320     | 5974.8909   |
 | 100000  | 59978.6810  | 59116.9250  |
 | 100000  | 58738.2830  | 59404.4569  |
 | 100000  | 60946.5689  | 58564.1380  |

mattip replied:
Did you try reversing the order of your benchmarking? When we tried this a few weeks ago, the results were that the two patterns take the same time, plus minus noise and measuring variations. I think we should try to use python operators as much as possible, it makes the code much more readable and conveys the message that NumPy performance can be achieved without needing special function calls.

khushi-411 replied:
Yes, I did implement and observed that the time taken by the program to run is reversed. 
Thanks for specifying the reason! I will change NumPy's functions to Python's functions.

rgommers replied:
This seems fixed now, I'll resolve it. If there is still anything that needs preserving for later, please unresolve.

seberg commented on benchmarks/python/optimized-numpy.py:
```diff
@@ -0,0 +1,123 @@
+    accelerations = np.zeros_like(positions)
+    accelerations1 = np.zeros_like(positions)
+
+    accelerations = compute_accelerations(accelerations, masses, positions)
```
Sorry, the work is done in-place, I had just not realized it at first.

khushi-411 commented on benchmarks/python/optimized-numpy.py:
```diff
@@ -0,0 +1,123 @@
+        vectors = position0 - positions[index_p0 + 1: nb_particles]
+
+        distances = np.square(vectors).sum(axis = 1)
+        coefs = distances ** 1.5
```
Yes, you are right. Thanks for your suggestion! 
I modified the code in the following ways:
- **The best one:** The time it took: `177.2893s`.
```
coefs = 1. / distances ** 1.5 
accelerations[index_p0] += np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis = 0) 
accelerations[index_p0 + 1: nb_particles] += (mass0 * vectors.T * coefs).T
```
 - The second best one: The time it took: `181.0388s`.
 ```
coefs = np.sqrt(distances ** 3)
_x = np.add(np.sum(
            np.divide(
                np.multiply(
                    masses[index_p0 + 1: nb_particles], -1 * vectors.T
                    ),
                coefs).T, axis = 0),
            accelerations[index_p0]
            )
accelerations[index_p0] = _x

_temp = np.add(
                np.divide(
                np.multiply(
                    mass0, vectors.T
                    ),
                coefs).T,
            accelerations[index_p0 + 1: nb_particles]
            )
accelerations[index_p0 + 1: nb_particles] = _temp
```
- The third one: The time it took: `188.0585s`.
```
coefs = distances ** 1.5
_x = np.add(np.sum(
            np.divide(
                np.multiply(
                    masses[index_p0 + 1: nb_particles], -1 * vectors.T
                    ),
                coefs).T, axis = 0),
            accelerations[index_p0]
            )
accelerations[index_p0] = _x

_temp = np.add(
                np.divide(
                np.multiply(
                    mass0, vectors.T
                    ),
                coefs).T,
            accelerations[index_p0 + 1: nb_particles]
            )
accelerations[index_p0 + 1: nb_particles] = _temp
```
- Not so good one: The time it took: `191.1135s`.
```
coefs = 1. / np.sqrt(distances**3
accelerations[index_p0] += np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis = 0) 
accelerations[index_p0 + 1: nb_particles] += (mass0 * vectors.T * coefs).T
```

These codes are run for `50` iterations for a dataset of size `16`. I'll update it in the PR. Thanks! 
I'm not sure whether it's okay to draw such conclusions for smaller datasets (here, `16`) and a lesser number of iterations (here, `50`).

khushi-411 replied:
Yes, thanks! I'll improve those codes.

seberg replied:
I just realized one possible point for confusing fluctuations (*possible* without profiling I have no idea if this matters at all):
```
In [1]: arr = np.random.random((1000, 3))
In [2]: arr2 = arr.copy("F")

In [3]: %timeit arr.sum(-1)
11.1 µs ± 466 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit arr2.sum(-1)
3.22 µs ± 3.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit arr[:, 0] + arr[:, 1] + arr[:, 2]
3.53 µs ± 49.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
The reason being that summing 3 elements can be slow in NumPy, NumPy is not smart enough to decide that 3 elements is so few that it would make sense to swap the iteration order (in most cases) – nor do we have blocked iteration that might also fix this.

(But I am not sure we should obfuscate the code even if this turns out to make a difference)

khushi-411 commented on benchmarks/python/optimized-numpy.py:
```diff
@@ -0,0 +1,123 @@
+    energy_previous = energy0
+
+    for step in range(nb_steps):
+        positions = sum(np.multiply(velocities, time_step), 0.5 * np.multiply(accelerations, time_step ** 2)) + positions
```
Ah, my bad. I realized it's not correct. I corrected the equation to the following eq:
```
positions += time_step*velocities + 0.5*accelerations*time_step**2
```

khushi-411 commented:
Hi,
I had addressed all the reviews and made those improvements. Thank you for all of your valuable words. Will you please let me know if any other corrections that have to be made? 

Few things are left to be updated:
- Python is too slow for an input of a size of `256`. It was taking too long for its execution, so I felt to update you all. I'll update the graph and result table once the code execution completes.
- I just now noticed the other typo in eq (II), position vectors. I'll correct it to `r_j = s_k - s_(k+1)`.

cc: @mattip @rgommers @seberg

rgommers commented on benchmarks/python/optimized-numpy.py:
```diff
@@ -0,0 +1,123 @@
+
+def compute_energies(masses, positions, velocities):
+
+    ke = 0.5 * (np.multiply(masses, np.square(velocities).sum(axis = 1)).sum())
```
This now reads `ke = 0.5 * masses @ np.sum(velocities**2, axis=1)`.

rgommers commented:
> And I included the performance for input sizes `32`, `64`, `128`, and `256`. Let me know if we have to change the values. Or is that okay?

That should be fine. The runtime is a bit long (up to 180 sec), but that's still doable to reproduce because we don't run these benchmarks a lot. I'd not go any larger than `256` though.

> Is that okay to use a non-vectorized method?

That should be fine. Sometimes it's even more readable. 

> Also, I expected the performance of Numba nearly equal to that of Pythran, but that's not the case here. What do you think about it?

You now use the code in `compiled_methods.py` for both Pythran and Numba, right? If so, it is slightly surprising that Numba is that much slower, but it can happen. On average they should be about as fast, but it can depend on a single line of code that's not well optimized if you're just looking at a single benchmark.

khushi-411 commented:
Hi, 
Thanks for your confirmations. 

> You now use the code in compiled_methods.py for both Pythran and Numba, right?

Yes, we are now using the same code for both Pythran and Numba.  

Numba's behavior is quite a bit interesting.

rgommers commented:
I think we should move the benchmarking code to a new repo, it doesn't quite fit here or in the main numpy repo. Proposed name: `numpyorg-benchmarks`. @mattip, @seberg, @melissawm any thoughts on that?

mattip commented:
I think something is off with the normalization. Originally, the formula `time / (n_particles**2)` was giving a almost constant time on all the implementations. Now it seems it is only constant for NumPy. If that is the case, we should rethink the normalization, something else is overshadowing the simple `n**2` model.

Moving the benchmark code elsewhere seems reasonable. Should I open a repo?

khushi-411 commented:
Hi, 
I agree with you. But at the same time, I am sure that the results are correct. It's just a way to look into things. Let's take an example of the two methods in the smaller area.

Let time(t) = `2 secs` and `n_particles = 2, 3, 4...`
- Using the formula `time / (n_particles)` to normalize: We'll get `2/2=1`, `2/3=0.67`, `2/4=0.5` etc.
- Using the formula `time / (n_particles**2)` to normalize: Here, we will get `2/4=0.5`, `2/(3x3)=0.22`, `2/(4x4)=0.125` etc. The rate of change differs.

The following is the output, in case we use `time / (n_particles**2)` for normalizing:
![performance_benchmarking](https://user-images.githubusercontent.com/62256509/135233668-7cb766c5-62f8-45c2-b0ad-2a6b4c4d1e96.png)

According to me, normalizing the results using the formula `time / (n_particles**2)` is a bit biased. Can we play with growth rates to present the benchmarking results?

What do you all think?

> Moving the benchmark code elsewhere seems reasonable. Should I open a repo?

We discussed transferring my repo https://github.com/khushi-411/numpyorg-benchmarks to NumPy org. Though, it needs some edits.
The structure I'm planning:
```
|-- data
    |-- all `input.txt` files here
|-- images
    |-- output images
|-- nbody problem
    |-- cpp
        |-- main.cpp
    |-- python
        |-- python_numpy.py
        |-- pure_python.py
        |-- compiled_methods.py
        |-- plot.py
    |-- README.md
        |-- describing the problem statement
|-- README.md
```
Is that okay?

Thanks!

cc: @mattip @rgommers @seberg @melissawm

mattip commented on content/en/benchmark.md:
```diff
@@ -0,0 +1,269 @@
+
+## Results
+
+Table values represent the normalized time taken in seconds by each algorithm to run on the given datasets for $50$ number of iterations. The raw timing data can be downloaded from <a href = "benchmarks/data/table.csv">here</a>.
```
Format to 80 columns, specify the normalization used in the table, and remove reduncancy
```suggestion
Table values represent the normalized time `time / n_particles` taken in
seconds by each algorithm to run on the given datasets for $50$ iterations. The
raw timing data can be downloaded from <a href =
"benchmarks/data/table.csv">here</a>.```

khushi-411 replied:
Sure, I'll make the edits. Thanks!

mattip commented:
> normalizing the results using the formula time / (n_particles**2) is a bit biased

Ahh, I was confused: you are currently normalizing by `time / n_particles`. Using `n_particles ** 2` seems to more-or-less flatten out the results for all the algorithms except NumPy, which gets faster for more particles (it would be nice to reason about this and add a comment why this happens. Hint: as we get more particles, the times move from "python-like" to "compiled-like"). What kind of a bias do you see? I can think of how to explain `n_particles ** 2` from the formulas, and it seems sufficient enough of a normalization for a first-order approximation. 

I think that is a more compelling story, perhaps the numerical text at the top of each column could present both the raw and the normalized values, as well as presenting both values in the table.

mattip commented:
I created  numpy/numpyorg-benchmarks. Please issue a PR to move the code there.

mattip commented:
Sorry, I created mattip/numpyorg-benchmarks instead. It seems I do not have the needed permissions to create a repo in the numpy org, and I didn't notice that github moved me to mattip instead. @rgommers could you create the repo?

rgommers commented:
> Sorry, I created mattip/numpyorg-benchmarks instead. It seems I do not have the needed permissions to create a repo in the numpy org, and I didn't notice that github moved me to mattip instead. @rgommers could you create the repo?

done!

khushi-411 commented:
I'm closing this PR because I shifted all the work to [numpyorg-benchmarks](https://github.com/numpy/numpyorg-benchmarks) repo. Thanks!
